### ISSUE TITLE
Direct file read/write access support for WebAssembly ?

### ISSUE THREAD

**Issue by agnivade** [WICG/native-file-system#139]
Hi,

WebAssembly allows us to run code written in native languages in the browser. Currently, file system access is not something that works in wasm because the browser did not have any such support so far. But now with this, things are changing.

One of the main issues with the current file system spec is that all reads and writes go through a file picker. That is problematic for wasm code which directly passes the file name to the `open` call in their native language, which can then be intercepted in the wasm interface.

I am aware of the security concerns. But this means file system support is still a no-go for WebAssembly. Are there any plans to look into this in the future ?

**Comment by mkruisselbrink** [comment]
To be clear, there is direct access to files/directories in a origin-scoped sandboxed filesystem via this API.

I'm also not sure how the limitation of having to go through a file picker to get access to files outside of that sandboxed file system matters for WebAssembly specifically? Other platforms (i.e. Android, UWP, probably others) have very similar limitations and requirements. You can access some "scoped" or "sandboxed" paths directly, and for other paths you need to go through a file picker/trusted UI. I don't see us changing that in the future either.

**Comment by agnivade** [comment]
> To be clear, there is direct access to files/directories in a origin-scoped sandboxed filesystem via this API.

Ah thank you, I overlooked it in the spec because it wasn't in the example. You are referring to this one right-  https://wicg.github.io/native-file-system/#api-getsystemdirectory ?

> I'm also not sure how the limitation of having to go through a file picker to get access to files outside of that sandboxed file system matters for WebAssembly specifically?

Because one can write `open("/home/user/file.txt")` and expect it to work in their own language. Then one can compile it to wasm and it works fine if run using Node.js. But then it doesn't work in the browser. This means if I have some code which reads some test data and from a file and works on it, and I want to test the code in wasm, I can't do that. This is effectively blocking us from running the entire test suite of the Go toolchain in the browser. Currently, we do that in Node. But there are some browser specific code paths (for eg. fetch), which does not get tested (and has therefore caused bugs to be caught by users). Of course, we can run the tests by skipping all the filesystem related ones. But that is extremely hacky and we would need to maintain the list of skipped tests separately.

Hope that makes sense.

**Comment by pwnall** [comment]
> > I'm also not sure how the limitation of having to go through a file picker to get access to files outside of that sandboxed file system matters for WebAssembly specifically?
> 
> Because one can write `open("/home/user/file.txt")` and expect it to work in their own language. Then one can compile it to wasm and it works fine if run using Node.js. But then it doesn't work in the browser. This means if I have some code which reads some test data and from a file and works on it, and I want to test the code in wasm, I can't do that. This is effectively blocking us from running the entire test suite of the Go toolchain in the browser. Currently, we do that in Node. But there are some browser specific code paths (for eg. fetch), which does not get tested (and has therefore caused bugs to be caught by users). Of course, we can run the tests by skipping all the filesystem related ones. But that is extremely hacky and we would need to maintain the list of skipped tests separately.
> 
> Hope that makes sense.

Thanks for bringing this up! I think we'll need a FAQ / disclaimer in the explainer, because I expect this will be a fairly common question.

Unfortunately, this confusion comes down to a fundamental difference in security models, and will be a rough edge for a while. Today, running a native app (whether it's compiled, perhaps by `go`, or interpreted, perhaps by `node`) implies granting it access to all the user's files. By contrast, users expect to be able to navigate to a web site without any side-effects, and we can only break this expectation with the user's explicit consent, usually via permission prompt.

For whatever it's worth, recent developments suggest that native apps will move towards the web model. For example, macOS 10.15 shows a permission prompt before allowing access to the user's Downloads directory. In the distant future, programmers may not be surprised that `open("/home/user/file.txt")` may involve a permission prompt.

For your concrete problem of running the Go test suite, you'll need to figure out how to bridge Go's expectation of a POSIX-like filesystem with the Web platform's model, where you can easily get full access to a sandboxed filesystem, but accessing the user's data involves file/directory handles that don't readily map to a tree. I suspect you'll make most progress in the short run by using the sandboxed filesystem.

I think the issue you've raised will come up in a lot of WASM scenarios. I hope you'll collaborate with the rest of the WASM community, and participate in proposing solutions. I'd also look for inspiration to mobile ports of various languages -- as far as I know, neither Android nor iOS let you open a user's files directly.

**Comment by agnivade** [comment]
Thanks ! Yep, I get the part about getting explicit consent from the user, and I can _probably_ work around it using some browser automation. The only issue that I have is that the API does not allow the code to specify a file path for which to ask the user for permission.

i.e., I should be able to specify that I want to open `/home/user/file.txt` and let the user accept it via a prompt.

I don't think it breaks any expectations, but it still allows the code to specify the file it wants to open.

What do you think about this ?

**Comment by pwnall** [comment]
> Thanks ! Yep, I get the part about getting explicit consent from the user, and I can _probably_ work around it using some browser automation. The only issue that I have is that the API does not allow the code to specify a file path for which to ask the user for permission.
> 
> i.e., I should be able to specify that I want to open `/home/user/file.txt` and let the user accept it via a prompt.

I think that should be discussed in its own issue. I quickly skimmed existing issues, and was surprised this doesn't exist.

> I don't think it breaks any expectations, but it still allows the code to specify the file it wants to open.

I don't think we're willing to let an app prompt the user for access to an arbitrary path. For example, I'm concerned that most users would not realize that a path pointing to the cookies database in their Chromium profile would compromise their accounts. 

> What do you think about this ?

If browser automation works for you, I recommend pursuing an addition to an automation protocol (WebDriver or DevTools). The addition would intercept a `chooseFileSystemEntries` call and cause the desired path to be selected, instead of invoking a file picker.

**Comment by agnivade** [comment]
> If browser automation works for you, I recommend pursuing an addition to an automation protocol (WebDriver or DevTools). The addition would intercept a chooseFileSystemEntries call and cause the desired path to be selected, instead of invoking a file picker.

Yep, I believe we already have something for that: https://chromedevtools.github.io/devtools-protocol/tot/Page#method-setInterceptFileChooserDialog. I think that suffices my requirements.

Let me know if you still want to discuss the new API in a separate issue, or it's a decided "no" from your end.

**Comment by oyiptong** [comment]
@agnivade, you mentioned running the go test suite.

If you're running unit tests, and you have control over the compilation output, you could consider using the sandboxed filesystem. It isn't a 100% match with the "native local" file system, but may work well for unit testing.

**Comment by agnivade** [comment]
Several unit tests read from testdata files and read/write to temp directories. So it's not possible to use a sandboxed filesystem.

**Comment by pwnall** [comment]
Apologies if I'm missing something... but would it be possible to create a minimal structure in the sandboxed filesystem before running the test suite? Based on the comment above, I expect you'd need a `tmp` directory and a copy of the go source tree.

**Comment by agnivade** [comment]
Oh you mean pre-create _every_ testdata file in the sandboxed filesystem, before starting the test ? That's one hell of a hack, but might just work :smile: . For now, as a first step, I am trying to intercept the file chooser dialog and supply the file path through automation as that is the simplest and the direct way forward for me.

**Comment by agnivade** [comment]
Wait, sorry let me take a step back. The sandboxed filesystem allows to read/write only to the sandbox. How am I going to read the testdata files from the filesystem, if I am to just work with the sandbox ?

**Comment by mkruisselbrink** [comment]
This feels like it is getting kind of off-topic, but one way you could read the testdata would be to just fetch it via whatever http server you're also serving the rest of the application.

I don't think there is much else actionable in this request, but please correct me if I'm wrong. There does seem to be overlap with #150, since having webdriver or something like that test automation for our tests might also help for your automated tests?

**Comment by agnivade** [comment]
> but one way you could read the testdata would be to just fetch it via whatever http server you're also serving the rest of the application.

That is indeed possible yes. But I was looking for a more direct approach.

We can close the issue for now. Thanks for following up.

**Comment by ivanjeremic** [comment]
> To be clear, there is direct access to files/directories in a origin-scoped sandboxed filesystem via this API.
> 
> I'm also not sure how the limitation of having to go through a file picker to get access to files outside of that sandboxed file system matters for WebAssembly specifically? Other platforms (i.e. Android, UWP, probably others) have very similar limitations and requirements. You can access some "scoped" or "sandboxed" paths directly, and for other paths you need to go through a file picker/trusted UI. I don't see us changing that in the future either.

Old issue but imagine someone uses WebAssembly to build a database? Maybe crazy idea but who knows.

**Comment by Blame4** [comment]
I used WASM to store file directly in /tmp, then memory increased. It seems the file system default in memory, not indb. Any way, you can always use indb even by c#.
